Fixes #13768 - random and MAC-based name generator

lzap opened the pull request:
This patch is a preview of name generator that replaces
https://github.com/theforeman/foreman/pull/3203. It is based on
foreman_name_generator gem which is currently a plugin and this is an attempt
to bring this into core. I'd probably rename the gem and strip down all the
Foreman related plugin stuff if this patch looks good

The gem contains frequently occurring given names and surnames from the 1990 US
Census (public domain data) which gives 33,554,432 (25 bits) total of male and
female name combinations. Built-in generator can either generate randomized
succession, or generate combinations based on MAC adresses.

Examples of MAC-based names:
- 24:a4:3c:ec:76:06 -> bobby-louie-sancher-weeler.my.lan
- 24:a4:3c:e3:d3:92 -> bob-louie-sancher-rimando.my.lan

MAC addresses with same OID part (24:a4:3c in this case) generates the same
middle name ("Louie Sancher" in the example above), therefore it is possible to
guess server (or NIC) vendor from it, or it should be possible to shorten
middle names (e.g. bobby-ls-weeler.my.lan) in homogeneous environments.

The random name generator makes use of [Fibonacci linear feedback shift
register](https://en.wikipedia.org/wiki/Linear_feedback_shift_register) which
gives deterministic sequence of pseudo-random numbers. Additionally, algorighm
makes sure names with same first name (or gender) and last name are not
returned in succession. Since there are about 1% of such cases, there are
about 33 million unique total names. During plugin installation, the register
is seeded with a random value, so each installation gets unique sequence.
Example sequence:
- velma-pratico.my.lan
- angie-warmbrod.my.lan
- grant-goodgine.my.lan
- alton-sieber.my.lan
- velma-vanbeek.my.lan
- don-otero.my.lan
- sam-hulan.my.lan

The polynomial used in linear feedback shift register is x^25 + x^24 + x^23 + x^22 + 1.

_Comparison of types_

MAC-based advantages
- reprovisioning the same server generates the same name
- middle names are same for unique hardware vendors

MAC-based disadvantages
- name is longer

Random-based advantages
- name is shorter

Random-based disadvantages
- reprovisioning the same server generates different name

Please do initial review and provide feedback, I will carry on with tests and
the gem rename and packaging after that.

lzap committed:
Fixes #13768 - random and MAC-based name generator

xprazak2 commented on app/models/setting/provisioning.rb:
```diff
@@ -23,9 +23,18 @@ def self.load_defaults
         self.set('dns_conflict_timeout', N_("Timeout for DNS conflict validation (in seconds)"), 3, N_('DNS conflict timeout')),
         self.set('clean_up_failed_deployment', N_("Foreman will delete virtual machine if provisioning script ends with non zero exit code"), true, N_('Clean up failed deployment')),
+        self.set('name_generator_type', N_("Type of name generator"), 'Off', N_("Name generator type")),
+        self.set('name_generator_seed', N_("Initial LFSR seed value (do not change)"), -1, N_("Name generator seed")),
```
If the setting is not supposed to be changed, the we should not allow users to change it. Maybe we need some kind of 'frozen' settings? But that is probably out of scope for this PR.

dLobatog replied:
:+1:  - I'd remove all options but the 'type' one. Here are few reasons:
- It's complicated (LFSR? most users won't know what it is nor they should need to know how the sausage is being made)
- User is given zero information why would they change that - I can't think of a good reason myself either
- It's yet another source of bugs

lzap replied:
I agree, I will just seed it. The question is - where should I put the seed value, if settings is not an option? Do I want to create new database table to hold one value? I think the proper solution is to implement read-only or even invisible settings.

dLobatog replied:
@lzap I think it's fine to do just like we do with the encryption key, put it in a module in an initializer (`NameGenerator::SEED`), or just in settings.yaml

lzap replied:
Sure, seed is not the issue, I am not sure about the register.

lzap replied:
For the next version, I am going with readonly settings. We already have support that for options which were overriden by settings.yaml.

domcleal replied:
Please don't abuse the settings system for storage of internal, global data/counters.

xprazak2 commented:
Works as expected for me, there is just chicken-and-egg problem related to MAC-based names. When provisioning to libvirt and creating a new host through UI, the MAC is generated after the form is submitted (unless typed in manually) and the form needs hostname to be valid.

dLobatog commented on app/models/host/base.rb:
```diff
@@ -102,6 +102,7 @@ def initialize(*args)
       super(*args)
 
       build_required_interfaces
+      self.name = NameGenerator.new.next_random_name if (self.name.nil? || self.name.empty?)
```
`self.name.nil? || self.name.empty?` is the same as `name.present?`

dLobatog commented on app/models/setting/provisioning.rb:
```diff
@@ -23,9 +23,18 @@ def self.load_defaults
         self.set('use_shortname_for_vms', N_("Foreman will use the short hostname instead of the FQDN for creating new virtual machines"), false, N_('Use short name for VMs')),
         self.set('dns_conflict_timeout', N_("Timeout for DNS conflict validation (in seconds)"), 3, N_('DNS conflict timeout')),
         self.set('clean_up_failed_deployment', N_("Foreman will delete virtual machine if provisioning script ends with non zero exit code"), true, N_('Clean up failed deployment')),
+        self.set('name_generator_type', N_("Type of name generator"), 'Off', N_("Name generator type")),
```
Why not default to 'Random'? The help message is a bit lacking on context.

lzap replied:
Fixed above comment. Sure, Random seems to be better default.

dLobatog commented on config/initializers/name_generator_seed.rb:
```diff
@@ -0,0 +1,6 @@
+if (Setting['name_generator_seed'] && Setting['name_generator_seed'] < 1) || (Setting['name_generator_register'] && Setting['name_generator_register'] < 1)
```
I'd say let's just ship a seed and register and forget about letting users change it

lzap replied:
Ok, I think I should move this to regular seed script rather than initializer.

dLobatog commented:
@lzap Thanks for the patch. Aside from the comments inline:
- When I set default to 'MAC+RANDOM' (why caps?) - and I click on New Host, I don't get the default hostname until I change the MAC. Is that expected? I thought when you select any option with 'Random', a random name is provided on 'New Host' automatically.
- I'd say MAC+RANDOM is a sane default, use MAC for discovered machines or when you want to set the MAC for some reason, but normally just use the random name. I'd argue few people set the MAC during provisioning - yet again we don't have data, just anecdotal observations.

dLobatog commented on Gemfile:
```diff
@@ -38,6 +38,8 @@ gem 'rails-observers', '~> 0.1'
 gem 'sprockets', '~> 3'
 gem 'sprockets-rails', '>= 2.3.3', '< 3'
+# FIXME This needs to be renamed if you like the patch
+gem 'foreman_name_generator', :path => '../foreman-name-generator'
```
Remove the path option, otherwise Jenkins will not be able to load it

lzap replied:
If there are no objections in having this in core, I will refactor the gem to be a simple gem instead of Foreman Plugin then and then change this line of course.

dLobatog replied:
Yeah, I think it's a good addition in general for Foreman & simple enough that making it a plugin defeats the purpose of providing sane defaults

lzap commented:
Getting back to this, I refactored and renamed the gem to https://github.com/lzap/deacon - it is no longer a Foreman Plugin.

lzap commented:
> Works as expected for me, there is just chicken-and-egg problem related to MAC-based names. When provisioning to libvirt and creating a new host through UI, the MAC is generated after the form is submitted (unless typed in manually) and the form needs hostname to be valid.

Yeah, in that case, it will fallback to random name. We can't do much about this.

lzap commented:
@dLobatog can you elaborate on your IRC mention of https://github.com/theforeman/foreman/blob/develop/config/initializers/encryption_key.rb.example please?

The thing is, this can be used for the seed value, but I can't use it for storing register which changes overe time.

dLobatog commented:
@lzap I'm certainly missing some background. Why would the register change?

lzap commented:
> @lzap I'm certainly missing some background. Why would the register change?

It's described here: https://github.com/lzap/deacon

The key thing is to store register (a number) and use it for each
generation in order to get non-repeating sequence of name combinations.

## 

Later,
 Lukas #lzap Zapletal

lzap commented:
> @lzap I'm certainly missing some background. Why would the register change?

I will introduce new table which will hold history of generated names
and its register, the last one can be read from it. Maybe I could use
Audit table for that, if I am able to store the number there along with
the audit record.

## 

Later,
 Lukas #lzap Zapletal

dLobatog commented:
@lzap ~~I'm a bit confused about the register thing. there are 33,554,432 combinations with a single register so why not keeping it as a constant like ENCRYPTION_KEY instead of a table? Why would an user change it?~~

Nevermind

lzap commented:
As an alternative to storing a register, simply random generator could be used. You made a good point that out of 33 million possible combinations, the probability of hitting existing host name is low. Still we need to check and the check is pretty expensive since names are stored in FQDN form, therefore database index cannot be used (at least through Rails API).

I think I should be able to take best out of both worlds - pick randomly generated number and use the sequence generator storing register in Rails cache until we hit an existing name. Then to purge the cache and generate another random number to start over. If Rails were restarted or cache wiped because of a previous hit, generate another random starting register. What you think @dLobatog ?

lzap commented:
@xprazak2 thanks for testing - MAC-based generator is only meant for bare-metal, where NIC MAC addresses does not change very often. For virtualized environments, MAC-based generator stability is not applicable, therefore in this case this should fallback to random based generator. If it doesn't, it's a bug.

I am currently working on storing the register in Rails cache, if it is not present, I will just generate a random value.

lzap commented:
I released `deacon` 1.0.0 gem, rebased current code. All the settings are gone, except the random generator type (MAC based or Randomized). The register is now stored in Rails cache, if the register is missing (e.g. initial use, or after upgrade), it is picked up randomly. How it works at the moment:

If you select MAC based names, you go to Interfaces and provide MAC of the primary interface, then name is autogenerated from the MAC. Both host name and interface domain name are filled in the moment you unfocus MAC field. If you do VM or Cloud provisioning, it does not make much sense to use MAC-based name generation (since MAC addresses are always random), therefore you should choose "Random-based" generator instead. If you attempt to use MAC-based anyway and MAC is missing during New Host creation, it falls back to Random-based.

For Random-based setting, the moment you enter New Host form, the randomized name is already pre-filled in the host name. If you visit the page once again (e.g. refresh), new random name is generated. I can perhaps implement an ad-hoc button "Randomize name" next to the host name so users are able to skip names they don't like (e.g. it's their boss name :-) or if they want to keep this option turned off and only do ad-hoc generation from time to time.

lzap commented:
The change is UI only, API should not be affected at all by this. Users can use own generators when using API or CLI, the underlaying gem could be used as well.

dLobatog commented on Gemfile:
```diff
@@ -38,6 +38,7 @@ gem 'protected_attributes', '~> 1.1.1'
 gem 'sprockets', '~> 3'
 gem 'sprockets-rails', '>= 2.3.3', '< 3'
 gem 'responders', '~> 2.0'
+gem 'deacon', '~> 1.0'
```
Just a reminder, we have to package this one

dLobatog commented on app/helpers/settings_helper.rb:
```diff
@@ -17,6 +17,12 @@ def value(setting)
         edit_select(setting, :value, {:select_values => Hash[Environment.all.map{|env| [env[:name], env[:name]]}].to_json } )
+      when /_generator_type$/
+        edit_select(setting, :value, {:select_values => {
+          'Off' => "Off",
```
These strings have to be translated

dLobatog commented:
@lzap :+1: Looks pretty good and nearly ready! I'm cool with merging this, just translate the strings in the helper & fix the tests and it'd be ready to roll on 1.12 :smile:

discreet commented:
out of curiosity would it be possible to use the UUID for the host name that foreman generates for puppet certificates (if you have the option enabled)? That would be ideal (in my scenario) because then your CA proxy browsing certificates matches up to the host name of your servers.

lzap commented:
> out of curiosity would it be possible to use the UUID for the host name that foreman generates for puppet certificates (if you have the option enabled)? That would be ideal (in my scenario) because then your CA proxy browsing certificates matches up to the host name of your servers.

So you want Foreman to generate a random UUID for the host name rather
than human readable name? Is this it? That should be indeed possible.
I can add that easily. This also reminds me we should name
"Random-based" differently, it should read "Random human name" instead I
think. Then we can add "Random UUID".

## 

Later,
 Lukas #lzap Zapletal

discreet commented:
@lzap yea. Right now I have the `use UUID for certs` option enabled which is perfect for my environment. It would be that much better if the random host name generator could use that same UUID (if the option is enabled) to set the hostname. My other thought was a regex people could set so it can randomly generate based on a naming convention but I don't like that as much.

lzap commented:
> @lzap yea. Right now I have the `use UUID for certs` option enabled which is perfect for my environment. It would be that much better if the random host name generator could use that same UUID (if the option is enabled) to set the hostname. My other thought was a regex people could set so it can randomly generate based on a naming convention but I don't like that as much.

I still don't get it, should it generate random UUID or reuse some other
one? I don't exactly understand what UUID are you talking about and what
is "use UUID for certs".

## 

Later,
 Lukas #lzap Zapletal

lzap commented:
I will implement UUID generator, it will set both `name` and `certname` so it matches Puppet cert hostname even if the setting is on. Will implement the button for ad-hoc generation and package the dependency next week.

dLobatog commented:
@lzap No need to do it in this PR though, please. It's nearly ready to be merged so you can safely make the changes in https://github.com/theforeman/foreman/pull/3240#issuecomment-215350685 , and implement the UUID in a branch based off this one.

lzap commented:
I would like to implement the ad-hoc button and turn this feature off by default, if you don't mind. It is a little bit intrusive IMHO.

lzap commented:
For the record, packages are being added under https://github.com/theforeman/foreman-packaging/pull/1122

lzap commented:
Extracted strings. Due to lack of time, I am not adding the UUID generator, but I created Randomize link next to the host name which is available for ad-hoc random name generation. Please re-review @dLobatog thanks!

timogoebel commented on app/assets/javascripts/host_edit_interfaces.js:
```diff
@@ -320,3 +320,19 @@ function update_fqdn() {
+    var mac = $('#interfaceModal').find('.interface_mac').val();
+    $.ajax({
+      type: "GET",
+      url: "/interfaces/random_name?mac=" + mac,
```
The url should probably not be hardcoded here... I think, it would be better to store it in a data attribute using rails magic and use the data in the js.

lzap replied:
Sure.

timogoebel commented on app/models/setting/provisioning.rb:
```diff
@@ -22,10 +28,17 @@ def self.load_defaults
     true
   end
+
+  def validate_name_generator_seed(record)
```
Where is this used?

lzap replied:
Leftover, good catch.

timogoebel commented on app/models/setting/provisioning.rb:
```diff
@@ -22,10 +28,17 @@ def self.load_defaults
         self.set('dns_conflict_timeout', N_("Timeout for DNS conflict validation (in seconds)"), 3, N_('DNS conflict timeout')),
-        self.set('clean_up_failed_deployment', N_("Foreman will delete virtual machine if provisioning script ends with non zero exit code"), true, N_('Clean up failed deployment'))
+        self.set('clean_up_failed_deployment', N_("Foreman will delete virtual machine if provisioning script ends with non zero exit code"), true, N_('Clean up failed deployment')),
+        self.set('name_generator_type', N_("Type of name generator"), 'Random-based', N_("Random gives unique names, MAC-based are longer but stable (and only works with bare-metal)"), nil, {:collection => Proc.new {GENERATOR_TYPES} }),
```
Name and description are mixed up. Could you change that please?

Also: Could you update the description by what you posted in this PR? Explaining why Mac based are better for bare metal etc.

lzap replied:
Mixed up?

I struggle to find a good short description, any ideas?

timogoebel replied:
... as discussed on IRC, how about:

``` ruby
self.set('name_generator_type', N_("Random gives unique names, MAC-based are longer but directly related to specific MAC (and only work with bare-metal)"), 'Random-based', N_("Type of name generator"), nil, {:collection => Proc.new {GENERATOR_TYPES} }),
```

lzap replied:
Ok done.

timogoebel commented on app/views/hosts/_form.html.erb:
```diff
@@ -27,8 +27,8 @@
-        <%= text_f f, :name, :size => "col-md-5", :value => name_field(@host),
-            :help_inline => _("This value is used also as the host's primary interface name.") %>
+        <%= text_f f, :name, :size => "col-md-4", :value => name_field(@host),
+            :help_inline => link_to_function(icon_text("random", _("Randomize")), "randomizeName()") + ' ' + _("This value is used also as the host's primary interface name.") %>
```
I think the link should be hidden when the name generator is off as it does not work then.
Could you place the link below the the input field as in the interface modal ("Suggest new" link)?

Maybe we can add a popover-modal that explains the used techniques (read: the different name generator settings). This is just an rough idea though and no requirement imho.

lzap replied:
Done, also it nows preselect and focus the name on load and on click so it's more usable. Added tooltip as well that explains how to disable this feature.

timogoebel commented:
@lzap : Looks good all in all. I left some comments inline though. 👍

lzap commented:
Updated, rebased.

timogoebel commented on app/models/setting/provisioning.rb:
```diff
@@ -1,5 +1,11 @@
+    'Off' => _('Off'),
+    'Random-based' => _('Random-based'),
+    'MAC-based' => _('MAC-based')
+  }
```
I think, it would be cleaner to store this in the name_generator service. You can then call `NameGenerator::Types` below.

timogoebel commented on app/helpers/hosts_helper.rb:
```diff
@@ -469,4 +469,10 @@ def multiple_proxy_select(form, proxy_feature)
   end
+
+  def randomize_mac_link
+    link_class = "hide" if Setting[:name_generator_type] != "Random-based"
```
How about `NameGenerator.random_based?` ?

timogoebel commented on app/services/name_generator.rb:
```diff
@@ -0,0 +1,37 @@
+  end
+
+  def next_mac_name(mac)
+    if @type =~ /^MAC/i && mac
```
How about extracting this to a (class) method, `if mac_based? && mac`, you could use `delegate :random_based?, to: self.class` to make this a little cleaner imho.

timogoebel commented:
@lzap: Just tested the ui. Works as expected. Good job. I left to minor comments. I think we can clean up the way the generator type is checked a little and then we're ready to go.

timogoebel commented on app/models/host/base.rb:
```diff
@@ -107,6 +107,7 @@ def initialize(*args)
       super(*args)
 
       build_required_interfaces
+      self.name = NameGenerator.new.next_random_name unless name.present?
```
Is this already tested anywhere? If not, please add a tiny test for this.

lzap commented:
All remarks done, also with an attempt to simplify Host::Base.initialize method for code climate (pulled out the list of attributes to constant).

timogoebel commented:
@lzap: Retested. Looks good. I'll merge when jenkins finished the tests. 💯

timogoebel commented:
@lzap : Please have a look at rubocop & test failures.

lzap commented:
Added missing permissions, fixed red test. I am dropping the `@type` member attribute as we moved the settings handling into the NameGenerator helper class.

lzap commented:
Please do not merge yet, under special circumstances (Discovery) I am getting:

```
Module::DelegationError: Host::Base#name= delegated to primary_interface.name=, but primary_interface is nil:
    app/models/host/base.rb:124:in `rescue in name='
    app/models/host/base.rb:124:in `name='
    app/models/concerns/interface_cloning.rb:14:in `block in setup_object_clone'
    app/models/concerns/interface_cloning.rb:13:in `each'
    app/models/concerns/interface_cloning.rb:13:in `setup_object_clone'
    app/models/concerns/interface_cloning.rb:6:in `setup_clone'
    app/models/host/base.rb:338:in `setup_clone'
    app/models/nic/managed.rb:109:in `setup_clone'
    app/models/concerns/orchestration.rb:76:in `valid?'
    test_after_commit (1.1.0) lib/test_after_commit/database_statements.rb:11:in `block in transaction'
    test_after_commit (1.1.0) lib/test_after_commit/database_statements.rb:5:in `transaction'
    app/models/concerns/foreman/sti.rb:29:in `save_with_type'
    test_after_commit (1.1.0) lib/test_after_commit/database_statements.rb:11:in `block in transaction'
    test_after_commit (1.1.0) lib/test_after_commit/database_statements.rb:5:in `transaction'
    test_after_commit (1.1.0) lib/test_after_commit/database_statements.rb:11:in `block in transaction'
    test_after_commit (1.1.0) lib/test_after_commit/database_statements.rb:5:in `transaction'
    app/models/host/base.rb:302:in `becomes'
    /home/lzap/work/foreman_discovery/app/services/foreman_discovery/host_converter.rb:7:in `to_managed'
    /home/lzap/work/foreman_discovery/test/unit/host_discovered_test.rb:208:in `block in <class:HostDiscoveredTest>'
```

It is caused by

```
self.name = NameGenerator.new.next_random_name
```

and I need to investigate tomorrow.

lzap commented:
Done, fixed the test and broke up the initializer for code climate.

lzap commented:
One more unit test failure fixed. There are also some upgrade test failures but I don't believe these are related. Let's see.

timogoebel commented:
> One more unit test failure fixed.

@lzap : Looks good, rubocop is still unhappy. Could you fix the complaints, please?

timogoebel commented:
@lzap : This now needs a rebase and some tiny rubocop fixes. I think, we're ready to merge then.

lzap commented:
Rebased, fixed! @timogoebel thanks.

timogoebel commented:
@lzap : This needs yet another rebase, sorry.

lzap commented:
Done.

timogoebel commented:
@lzap : Thanks. I'll merge when jenkins seconds this.

timogoebel commented:
Merged as 0351034151a70be5bac559065c58056801bb3b22, thanks @lzap!
